### PR TITLE
Add support for arm64 RPI cross compile in build_examples.py

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Compiling for RPi on the vscode image is faster than trying to use rpi itself

#### Change overview
Add arm64 as a cross compile target for the linux apps

#### Testing

```
./scripts/build/build_examples.py --platform host --board arm64 build
```

built 3 apps (chip-tool, thermostat and all-cluster-apps examples)

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: scripts/build/builders/host.py [hunk 1]
```diff
@@ -45,21 +46,61 @@ def BinaryName(self):
...
+
+    # cross-compile support
+    CROSS_COMPILE_ARM64 = auto()
+
+    def PlatformName(self):
+        if self == HostBoard.NATIVE:
+            uname_result = uname()
+            return '-'.join([uname_result.system.lower(), uname_result.machine])
+        elif self == HostBoard.CROSS_COMPILE_ARM64:
+            return 'linux-rpi'
```
Comment (mspang) [inline scripts/build/builders/host.py:60]: linux-arm64

File: scripts/build/builders/host.py [hunk 2]
```diff
@@ -45,21 +46,61 @@ def BinaryName(self):
...
+class HostBoard(Enum):
+    NATIVE = auto()
+
+    # cross-compile support
+    CROSS_COMPILE_ARM64 = auto()
+
+    def PlatformName(self):
+        if self == HostBoard.NATIVE:
+            uname_result = uname()
+            return '-'.join([uname_result.system.lower(), uname_result.machine])
```
Comment (mspang) [inline scripts/build/builders/host.py:58]: Please check machine, if it's x86_64, label it something like linux-x64. Otherwise bail.

The name should be the same as the name used when cross compiling an x64 binary from the Pi. Compiling is symmetrical; it's is _how_ you build not _what_ you are building. This label should be the 'what' and it should ignore the 'how'.

File: scripts/build/builders/host.py [hunk 3]
```diff
@@ -45,21 +46,61 @@ def BinaryName(self):
...
+            ]
+        else:
+            raise Exception('Unknown host board type: %r' % self)
+
+    def GnBuildEnv(self):
+        if self.board == HostBoard.NATIVE:
+            return None
+        elif self.board == HostBoard.CROSS_COMPILE_ARM64:
+            return {
+                'PKG_CONFIG_PATH': self.SysRootPath('SYSROOT_AARCH64') + '/lib/aarch64-linux-gnu/pkgconfig',
```
Comment (mspang) [inline scripts/build/builders/host.py:95]: Why is this needed?

**Review by woody-apple — APPROVED** [review]

File: scripts/build/builders/host.py (hunk 2 above)
Comment (andy31415) [reply scripts/build/builders/host.py]: Updated.  In theory this may also compile 'darwin'so I left code logic a bit different. I allow cross compile if the 'native' platform is not the same as the cross compile.

File: scripts/build/builders/host.py (hunk 3 above)
Comment (andy31415) [reply scripts/build/builders/host.py:105]:
```
2021-09-23 14:47:39 INFO    Package avahi-client was not found in the pkg-config search path.
2021-09-23 14:47:39 INFO    Perhaps you should add the directory containing `avahi-client.pc'
2021-09-23 14:47:39 INFO    to the PKG_CONFIG_PATH environment variable
2021-09-23 14:47:39 INFO    No package 'avahi-client' found
2021-09-23 14:47:39 INFO    Traceback (most recent call last):
2021-09-23 14:47:39 INFO      File "/workspaces/connectedhomeip/examples/all-clusters-app/linux/third_party/connectedhomeip/build/config/linux/pkg-config.py", line 259, in <module>
2021-09-23 14:47:39 INFO        sys.exit(main())
Traceback (most recent call last):
2021-09-23 14:47:39 INFO      File "/workspaces/connectedhomeip/examples/all-clusters-app/linux/third_party/connectedhomeip/build/config/linux/pkg-config.py", line 156, in main
2021-09-23 14:47:39 INFO        prefix = GetPkgConfigPrefixToStrip(options, args)
2021-09-23 14:47:39 INFO      File "/workspaces/connectedhomeip/examples/all-clusters-app/linux/third_party/connectedhomeip/build/config/linux/pkg-config.py", line 100, in GetPkgConfigPrefixToStrip
2021-09-23 14:47:39 INFO        prefix = subprocess.check_output([options.pkg_config,
2021-09-23 14:47:39 INFO      File "/usr/lib/python3.9/subprocess.py", line 424, in check_output
2021-09-23 14:47:39 INFO        return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
2021-09-23 14:47:39 INFO      File "/usr/lib/python3.9/subprocess.py", line 528, in run
2021-09-23 14:47:39 INFO        raise CalledProcessError(retcode, process.args,
2021-09-23 14:47:39 INFO    subprocess.CalledProcessError: Command '['pkg-config', '--variable=prefix', 'avahi-client']' returned non-zero exit status 1.
2021-09-23 14:47:39 INFO    
2021-09-23 14:47:39 INFO    See //third_party/connectedhomeip/src/platform/BUILD.gn:34:3: whence it was called.
2021-09-23 14:47:39 INFO      pkg_config("avahi_client_config") {
2021-09-23 14:47:39 INFO      ^----------------------------------
2021-09-23 14:47:39 INFO    See //third_party/connectedhomeip/src/lib/BUILD.gn:30:5: which caused the file to be included.
2021-09-23 14:47:39 INFO        "${chip_root}/src/platform",
2021-09-23 14:47:39 INFO        ^--------------------------
```

File: scripts/build/builders/host.py [hunk 4]
```diff
@@ -45,21 +46,71 @@ def BinaryName(self):
...
 
 
-def ConcretePlatformName():
-    uname_result = uname()
-    return '-'.join([uname_result.system.lower(), release(), uname_result.machine])
+class HostBoard(Enum):
+    NATIVE = auto()
+
+    # cross-compile support
+    CROSS_COMPILE_ARM64 = auto()
```
Comment (mspang) [inline scripts/build/builders/host.py:53]: Just ARM64 please.

File: scripts/build/builders/host.py [hunk 5]
```diff
@@ -45,21 +46,71 @@ def BinaryName(self):
             raise Exception('Unknown app type: %r' % self)
 
 
-def ConcretePlatformName():
-    uname_result = uname()
-    return '-'.join([uname_result.system.lower(), release(), uname_result.machine])
+class HostBoard(Enum):
+    NATIVE = auto()
```
Comment (mspang) [inline scripts/build/builders/host.py:50]: This should likely just name x64.

The problem is that if your host is ARM64 these two boards are the same, so when you go to build "all boards" you build the same thing twice.

I don't now how to handle --board host, maybe an alias?

Comment (andy31415) [reply scripts/build/builders/host.py:50]: It is an awkwardness because both 'platform host' and 'board native' mean the same thing. All other platforms make a tuple, this one stutters as the same info is duplicated.

Maybe placing the cross compile into platform was wrong ... will try to rethink.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: Updated logic a bit, now everything looks like `linux-x64-chip-tool` or `linux-arm64-all-clusters-app`.

Comment (mspang) [reply scripts/build/builders/host.py:50]: Still wondering what happens if you build 'all boards' on the Pi.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: I can try it out, but then arm64 will be filtered out and only the 'native' one will take over due to the condition:

```py
if HostBoard.NATIVE.BoardName() != HostBoard.ARM64.BoardName():
    _MATCHERS[Platform.HOST].AcceptBoard(Board.ARM64, board=HostBoard.ARM64)
```

The RPI will not accept the arm64 board, only native and native will be converted to arm64.

Comment (mspang) [reply scripts/build/builders/host.py:50]: I think it is a bug to only accept native. We should be able to paste command lines between host platforms and get the same results.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: I would prefer to do that as a followup. Native was a convenient way to say 'build whatever is appropriate to run chip-tool locally'

This can be done in defaults for sure, however it seems to move the intent of this PR (I want cross compile) a bit down the rabbit hole.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: Probably aliasing should be done, so that both 'arm64' and 'native' mean the same thing. This would mean I would also need to add x64 at least (don't think we have/need x86 currently) and logic for removing aliasing issues.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: I am not yet in favor of dropping native/host. It seems convenient to tell people to run `build_examples.py --platform host --app chip-tool` to build chip tool and not worry about what platform they are on.

Comment (andy31415) [reply scripts/build/builders/host.py:50]: Tested on RPi (took a long time to bootstrap). Current PR version does what I would expect

```
ubuntu@ubuntu:~/connectedhomeip$ ./scripts/build/build_examples.py --platform host gen
2021-09-23 16:10:22 INFO    Platforms being built: host
2021-09-23 16:10:22 INFO    Boards being built: native
2021-09-23 16:10:22 INFO    Applications being built: all-clusters, chip-tool, thermostat
2021-09-23 16:10:22 INFO    Generating /home/ubuntu/connectedhomeip/out/linux-arm64-all_clusters
2021-09-23 16:10:22 INFO    Generating linux-arm64-all_clusters
2021-09-23 16:10:23 INFO    Done. Made 82 targets from 92 files in 792ms
2021-09-23 16:10:23 INFO    Command ['gn', 'gen', '--check', '--fail-on-unused-args', '--root=/home/ubuntu/connectedhomeip/examples/all-clusters-app/linux', '/home/ubuntu/connectedhomeip/out/linux-arm64-all_clusters'] completed
2021-09-23 16:10:23 INFO    Generating /home/ubuntu/connectedhomeip/out/linux-arm64-chip_tool
2021-09-23 16:10:23 INFO    Generating linux-arm64-chip_tool
2021-09-23 16:10:23 INFO    Done. Made 76 targets from 88 files in 770ms
2021-09-23 16:10:23 INFO    Command ['gn', 'gen', '--check', '--fail-on-unused-args', '--root=/home/ubuntu/connectedhomeip/examples/chip-tool', '/home/ubuntu/connectedhomeip/out/linux-arm64-chip_tool'] completed
2021-09-23 16:10:23 INFO    Generating /home/ubuntu/connectedhomeip/out/linux-arm64-thermostat
2021-09-23 16:10:23 INFO    Generating linux-arm64-thermostat
ls o2021-09-23 16:10:24 INFO    Done. Made 82 targets from 92 files in 776ms
2021-09-23 16:10:24 INFO    Command ['gn', 'gen', '--check', '--fail-on-unused-args', '--root=/home/ubuntu/connectedhomeip/examples/thermostat/linux', '/home/ubuntu/connectedhomeip/out/linux-arm64-thermostat'] completed
uubuntu@ubuntu:~/connectedhomeip$ ls out
linux-arm64-all_clusters  linux-arm64-chip_tool  linux-arm64-thermostat
```

Comment (andy31415) [reply scripts/build/builders/host.py:50]: Also see pains in the way we do unit testing: we generate a 'dry run' and output commands.

But now we start having a difference between what is available in one platform vs others, so our unit test will fail on all arm64 platforms. I am not happy with where this is going, would rather look at it in following changes.

File: scripts/build/builders/host.py (hunk 3 above)
Comment (mspang) [reply scripts/build/builders/host.py:112]: Ok, let's figure that out.. The idea with 'sysroot' param is that's all you need, it's broken if you also need an environment variable.

**Review by mspang — APPROVED** [review]
